app/screens/select_server/select_server.js

```diff
@@ -167,27 +150,87 @@ class SelectServer extends PureComponent {
     onClick = wrapWithPreventDoubleTap(async () => {
         const preUrl = urlParse(this.props.serverUrl, true);
         const url = stripTrailingSlashes(preUrl.protocol + '//' + preUrl.host);
-        let error = null;
 
         Keyboard.dismiss();
 
-        if (isValidUrl(url)) {
-            Client4.setUrl(url);
-            Client.setUrl(url);
-            this.props.actions.handleServerUrlChanged(url);
-            await this.props.actions.getPing();
-        } else {
-            error = {
-                intl: {
-                    id: 'mobile.server_url.invalid_format',
-                    defaultMessage: 'URL must start with http:// or https://'
+        if (this.state.connecting) {
+            this.cancelPing();
+
+            return;
+        }
+
+        if (!isValidUrl(url)) {
+            this.setState({
+                error: {
+                    intl: {
+                        id: 'mobile.server_url.invalid_format',
+                        defaultMessage: 'URL must start with http:// or https://'
+                    }
                 }
-            };
+            });
+
+            return;
         }
 
-        this.setState({error});
+        this.pingServer(url);
     });
 
+    pingServer = (url) => {
+        this.setState({
+            connecting: true,
+            error: null
+        });
+
+        Client4.setUrl(url);
+        Client.setUrl(url);
+        this.props.actions.handleServerUrlChanged(url);
+
+        let cancel = false;
+        this.cancelPing = () => {
+            cancel = true;
+
+            this.setState({
+                connecting: false
+            });
+
+            this.cancelPing = null;
+        };
+
+        this.props.actions.getPing().then((result) => {
+            if (cancel) {
+                return;
+            }
+
+            this.setState({
+                connecting: false,
+                error: result.error
+            });
+
+            if (!result.error) {
+                if (Config.EnableMobileClientUpgrade) {
+                    this.props.actions.setLastUpgradeCheck();
```

Comment:
I think you need to make sure that by this point you actually have config and license set in the store or else further screens will fail

Reply:
Agreed, eventually for the client upgrade we'll actually need the server config since the local config values will eventually be migrated to the server.

Reply:
Oh, I thought that was done as part of the getPing call. I didn't realize it was done in the background by something else. I'll look into it some more